index.bs

```diff
@@ -447,16 +490,30 @@ partial interface Navigator {
 
 ## ML ## {#api-ml}
 <script type=idl>
+enum MLDevicePreference {
+  // Let the user agent decide the most suitable device to use.
+  "default",
+  // Prefer a graphics processor or GPU.
+  "graphics",
+  // Prefer a compute-specific processor such as a neural engine or NPU.
+  "compute",
+  // Prefer a traditional software-based device such as a CPU.
+  "software"
```

Comment:
I prefer to naming it as `cpu`. It would be a clear device name that the Wasm libs could select to avoid the unnecessary data copying cross devices, like the [Support CPU - WebAssembly scenario of the op level execution use case](https://github.com/webmachinelearning/webnn/issues/156) describes.

Reply:
Other than the API name prefix, most API naming conventions generally discourage the use of abbreviations as identifier names. If we were to call it "cpu" we would also need a "gpu" value to be consistent, and then what's for AI accelerators? "npu", "vpu", "tpu"? I think that would start a naming discussion to no end.

Reply:
I suppose the "cpu" and "gpu" abbreviations would be well known by developers because they are already widely used by native ML APIs, e.g. NNAPI [`DeviceTypeCode`](https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1ga61f16b495d2e04593c819c12e42ac9f2) defines `ANEURALNETWORKS_DEVICE_CPU` and `ANEURALNETWORKS_DEVICE_GPU`, ML Compute [`MLCDeviceType`](https://developer.apple.com/documentation/mlcompute/mlcdevicetype) defines `gpu` and `cpu` and oneDNN [`dnnl_engine_kind_t`](https://oneapi-src.github.io/oneDNN/group__dnnl__api__engine.html#ga04b3dd9eba628ea02218a52c4c4363a2) also defines `dnnl_cpu` and `dnnl_gpu`.

For AI accelerators, probably we can just call them as "accelerator" (in the context of WebNN). This is also used by native APIs, such as NNAPI [`DeviceTypeCode`](https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1ga61f16b495d2e04593c819c12e42ac9f2) defines `ANEURALNETWORKS_DEVICE_ACCELERATOR` for "Dedicated accelerator for Machine Learning workloads.". Another example is OpenCL [`cl_device_type`](https://www.khronos.org/registry/OpenCL/specs/2.2/html/OpenCL_API.html#CL_DEVICE_TYPE) defines `CL_​DEVICE_​TYPE_​CPU`, `CL_DEVICE_TYPE_GPU` and `CL_​DEVICE_​TYPE_​ACCELERATOR` for heterogeneous hardware.

Actually, today's modern CPUs already have ML specific instructions, such as Vector Neural Network Instruction (VNNI) of X86 and Scalable Vector Extension 2 (SVE2) of Arm, and are adding even more advanced instructions, e.g., for matrix multiplication. My major concern is "software" seems not to reflect these new capabilities that would accelerate the ML workloads for web apps.

Reply:
I'm open for naming suggestion if others have input on this. In my line of work designing Windows API at Microsoft, we generally stay away from using abbreviations because its meaning and perceived freshness could change over time. (Windows API is well known for longevity).

Reply:
My 2 cents worth: "compute" and "software" sound too generic, and I wouldn't be able to guess what they meant.  On the other hand, I do think most people can guess what "cpu" or "gpu" mean. I agree custom neural accelerators don't have as obvious a choice. "npu" or "accelerator" or "custom" sound fine to me.

Reply:
+1 on cpu and gpu

Reply:
I couldn't find any reference or recommendations regarding usage of abbreviations in the Web API definition. Anyone knows where to look? @anssiko. If it is allowed, should it be used in lowercase or uppercase? etc.

DirectX API in general avoids abbreviations, and terms like "graphics" or even "software" have been used for quite some time. e.g. 

https://docs.microsoft.com/en-us/windows/win32/direct3d11/atoc-dx-graphics-direct3d-11
https://docs.microsoft.com/en-us/windows/win32/api/d3dcommon/ne-d3dcommon-d3d_driver_type

Reply:
@wchao1115 Web Platform Design Principles > Naming principles https://w3ctag.github.io/design-principles/#naming-is-hard is probably the best resource. Naming is always a topic of active discussion in API design, and there's often no absolute rights or wrongs.

Given there's a Web API called WebGPU, I suspect "gpu" would pass if you prefer it that way. Similarly, we can probably agree that "cpu" is familar to web developers of today.

That said, I'd challenge the group to think this from the perspective of future-proofing. How do the possible future architectures fit into the device buckets we choose? We could make this extensible with new buckets in the future, or we could make the initial ones abstract enough to accommodate future devices. I think there's always a trade-off whatever we choose. What I think we want to achieve is to be able to provide web developers with knobs they can use to give hints to the implementation so their workloads can be executed in a manner that yields the best user experience.

Reply:
Thanks @anssiko. I'll take this advice in the design principles: *Please consult widely on names in your APIs.* 

The hardest is probably the accelerator type. Let's discuss this naming in our next CG call.